Add a more comprehensive entire bidder timeout API

MattMenke2 opened the pull request:
Here's a proposal to add a timeout to limit the entire time each buyer takes to generate bids.  A single timer covers all of a bidder's interest groups.  The timer is based strictly on wall clock time, includes starting processes and downloading resources, and doesn't stop when, e.g., no buyer scripts are running due to CPU contention or whatnot.

Does this meets folks' needs for a more comprehensive bidder timeouts?

jeffkaufman commented on FLEDGE.md:
```diff
@@ -202,6 +206,8 @@ The returned `auctionResultPromise` object is _opaque_: it is not possible for a
 
 Optionally, `sellerTimeout` can be specified to restrict the runtime (in milliseconds) of the seller's `scoreAd()` script, and `perBuyerTimeouts` can be specified to restrict the runtime (in milliseconds) of particular buyer's `generateBid()` scripts. If no value is specified for the seller or a particular buyer, a default timeout of 50 ms will be selected. Any timeout higher than 500 ms will be  ...
 
+`perBuyerBidGenerationTimeouts` is structured like `perBuyerTimeouts`, but the values cover the entirety of the time it takes to generate bids for all interest groups for each buyer, including downloading resources, starting processes, and all generate bid calls. The single limit applies collectively across all interest groups with the same owner. It's measure as wall clock time starting when ded ...
```
effot -> effort

MattMenke2 replied:
Thanks, done (and found yet another typo as well...)

JensenPaul commented on FLEDGE.md:
```diff
@@ -202,6 +206,8 @@ The returned `auctionResultPromise` object is _opaque_: it is not possible for a
 
 Optionally, `sellerTimeout` can be specified to restrict the runtime (in milliseconds) of the seller's `scoreAd()` script, and `perBuyerTimeouts` can be specified to restrict the runtime (in milliseconds) of particular buyer's `generateBid()` scripts. If no value is specified for the seller or a particular buyer, a default timeout of 50 ms will be selected. Any timeout higher than 500 ms will be  ...
 
+`perBuyerBidGenerationTimeouts` is structured like `perBuyerTimeouts`, but the values cover the entirety of the time it takes to generate bids for all interest groups for each buyer, including downloading resources, starting processes, and all generate bid calls. The single limit applies collectively across all interest groups with the same owner. It's measure as wall clock time starting when ded ...
```
```suggestion
Optionally, `perBuyerBidGenerationTimeouts` is structured like `perBuyerTimeouts`, but the values cover the entirety of the time it takes to generate bids for all interest groups for each buyer, including downloading resources, starting processes, and all generate bid calls. The single limit applies collectively across all interest groups with the same owner. It's measured as wall clock time starting when dedicated tasks needed to generate a bid for a particular buyer start. Timers may be running for multiple bidders simultaneously.  It does not include time taken by the seller to score the buyer's bids. Once the timer expires, the affected buyer's interest groups may no longer generate any bids. Scripts may be unloaded, fetches cancelled, etc. All bids generated before the timeout will continue to participate in the auction. FLEDGE implementations should attempt, on a best-effort basis, to generate bids for each buyer in priority order, so lower priority interest groups are the ones more likely to be timed out.
```

sbelov commented on FLEDGE.md:
```diff
@@ -179,6 +179,10 @@ const myAuctionConfig = {
                         'https://www.another-buyer.com': 200,
                         '*': 150,
                         ...},
+  'perBuyerBidGenerationTimeouts': {'https://www.example-dsp.com': 500,
```
nit: the naming, admittedly, seems confusing. the current `perBuyerTimeouts` control the maximum runtime of each individual `generateBid` function invocation, and `perBuyerBidGenerationTimeouts` could be interpreted as doing exactly that – controlling the maximum runtime of the individual `generateBid` function.

Perhaps, the new field could be named `perBuyerOverallBiddingTimeouts` or `perBuyerMaxBiddingRuntime`.

MattMenke2 replied:
Went with perBuyerOverallBiddingTimeouts for now, to keep the "timeout" terminology consistent.

sbelov commented on FLEDGE.md:
```diff
@@ -179,6 +179,10 @@ const myAuctionConfig = {
                         'https://www.another-buyer.com': 200,
                         '*': 150,
                         ...},
+  'perBuyerBidGenerationTimeouts': {'https://www.example-dsp.com': 500,
```
Another API-ness nit: the auction configuration already contains a number of per-buyer dictionaries. I wonder if it would make sense to add some form of a buyer-level configuration object (which may include various timeouts and other constraints, as well as per-buyer signals), and then add a single per-buyer-origin dictionary, with values containing such a buyer-level configuration object. I understand there's a backwards compatibility issue if this route were taken, though FLEDGE may be in an early enough testing phase to be able to afford the change.

MattMenke2 replied:
We've been thinking that we might make just that change before standardization, so as not to break folks during field trials/experimentation.

sbelov commented on FLEDGE.md:
```diff
@@ -202,6 +206,8 @@ The returned `auctionResultPromise` object is _opaque_: it is not possible for a
 
 Optionally, `sellerTimeout` can be specified to restrict the runtime (in milliseconds) of the seller's `scoreAd()` script, and `perBuyerTimeouts` can be specified to restrict the runtime (in milliseconds) of particular buyer's `generateBid()` scripts. If no value is specified for the seller or a particular buyer, a default timeout of 50 ms will be selected. Any timeout higher than 500 ms will be  ...
 
+`perBuyerBidGenerationTimeouts` is structured like `perBuyerTimeouts`, but the values cover the entirety of the time it takes to generate bids for all interest groups for each buyer, including downloading resources, starting processes, and all generate bid calls. The single limit applies collectively across all interest groups with the same owner. It's measure as wall clock time starting when ded ...
```
> FLEDGE implementations should attempt, on a best-effort basis, to generate bids for each buyer in priority order

In the case of end user devices with a limited number of hardware threads, is there a recommendation on how FLEDGE implementations should schedule the CPU execution time _among_ individual buyers, given `perBuyerBidGenerationTimeouts` constraints are present and measure the wall clock time? For example, should the bidding process for different interest group buyers happen sequentially or in parallel (perhaps, relying on OS-provided scheduling), and how would that interact with the measurement of `perBuyerBidGenerationTimeouts`?

MattMenke2 replied:
> > FLEDGE implementations should attempt, on a best-effort basis, to generate bids for each buyer in priority order
> 
> In the case of end user devices with a limited number of hardware threads, is there a recommendation on how FLEDGE implementations should schedule the CPU execution time _among_ individual buyers, given `perBuyerBidGenerationTimeouts` constraints are present and measure the wall clock time? For example, should the bidding process for different interest group buyers happen sequentially or in parallel (perhaps, relying on OS-provided scheduling), and how would that interact with the measurement of `perBuyerBidGenerationTimeouts`?

Good question!

So Chrome for desktop's implementation loads and runs scripts for 10 buyers and 3 sellers at a time (which is arbitrary - number is more based on concerns about process overhead than CPU contention), loading buyers in the order they appear in the auction.

I'm not sure what, if anything, we'd want to add about specific details, though CPU contention and implementation details certainly come into consideration here.  I think it makes sense to recommend buyers be loaded in the order they appear in the auction - we generally want to put as much control in the hands of auction participants as we reasonably can.  Chrome does this, though if there's a FLEDGE process already running for a buyer, the buyer will jump the queue.  On Android, where we don't use a dedicated FLEDGE process, this behavior currently means all buyers typically run in parallel.

There's a tradeoff between speccing things out in more detail to force consistency across browsers/platforms in the spec, and allow browsers to optimize behavior given their own resource constraints here.

The wording here sort of assumes that scripts are loaded and then run on a per-buyer basis immediately.  Could imagine a browser implementation that downloaded all scripts at once, and then scheduled them based on CPU availability, in which case, these would effectively become timeouts for the entire bid generation phase of the auction.

sbelov replied:
> loading buyers in the order they appear in the auction

If some form of a queue-like mechanism was employed by a browser due to resource constraints (such as # of threads or processes that can execute in parallel), would a timer for the given buyer only be started once that buyer's task(s) are removed from the start of the queue, and the buyer's `generateBid` functions received a chance to execute (as opposed to the time when these tasks had originally been added to the queue)?

MattMenke2 replied:
> > loading buyers in the order they appear in the auction
> 
> If some form of a queue-like mechanism was employed by a browser due to resource constraints (such as # of threads or processes that can execute in parallel), would a timer for the given buyer only be started once that buyer's task(s) are removed from the start of the queue, and the buyer's `generateBid` functions received a chance to execute (as opposed to the time when these tasks had originally been added to the queue)?

The intention is that it would start as soon as something is done specifically for the buyer that delays the auction, such as starting to download the JS/JSON (In Chrome's case, this would actually be starting the process to download the JS/JSON - Chrome then runs the JS as soon as it has downloaded all it needs, modulo platform CPU time queuing).

The intention is to try and keep this simple and allow for future speed improvements (e.g., if we had an API that set a time limit on the combined JS execution, but someone had a 32 core CPU and run a bunch of bidder scripts in parallel, odds are that it's wall clock time that matters to sellers, not the total time across all 32/64 threads all running bidder scripts at once)

fhoering commented:
How would buyer latency reporting be handled in this case ? 

A buyer would need to know which interest groups constantly time out for example, if the problem is the trusted server call or the bidding script, if too many interest groups have been created for one user, .. Currently a buyer can only monitor the trusted server calls but as this will move to a trusted execution environment it seems best that Chrome also does reporting for those calls from the client side.

Related ticket but which only asks for seller latency reporting: https://github.com/WICG/turtledove/issues/299.

MattMenke2 commented:
> How would buyer latency reporting be handled in this case ?
> 
> A buyer would need to know which interest groups constantly time out for example, if the problem is the trusted server call or the bidding script, if too many interest groups have been created for one user, .. Currently a buyer can only monitor the trusted server calls but as this will move to a trusted execution environment it seems best that Chrome also does reporting for those calls from the client side.
> 
> Related ticket but which only asks for seller latency reporting: #299.

I'm not sure - there's no currently proposed declarative reporting API for bidders or sellers (In the case they can't run Javascript - e.g., if a buyer is slow and makes no bids, or a buyer fails to load, etc), and a bunch of things that could likely use one.

Covidnining reviewed (approved):
Link

MattMenke2 commented:
> Hi there, any update on the 'perBuyerBidGenerationTimeouts' change?

We ended up going with https://github.com/WICG/turtledove/pull/342 instead.  That doesn't preclude implementing this as well, of course.  This was designed as an alternative to the abort API, since aborting doesn't support a putative future world in which we don't reveal auction success/failure/completion time to the page running the auction, but until/unless we make changes along those lines, an abort API lets pages implement a timeout themselves instead.

MattMenke2 commented:
Just a heads up to anyone who's been following the PR - we're looking at implementing this again.

I've updated the PR to be on top of the latest version, and added mention of how this works when fields in the auctionConfig are promises.  I'm not sure if which auctionConfig values can be promises has been documented yet, but the timer only starts when all promises in the config blocking a particular buyer's generateBid() call have been resolved (since we're waiting on the seller, not the bidder).

That does potentially give buyers, at least those we load first, some extra free time to load resources, in the case an auctionConfig has promises blocking the buyer scripts from running (but not their scripts or signal files from loading).  We could alternatively start the timer when we start loading buyer scripts, pause it once those are loaded and we're waiting on the seller, but if a buyer has multiple scripts (and signals files), things get a bit weird.  Seems simplest to just count only time where the buyer is potentially blocking the auction from completing.

dmdabbs commented:
>That does potentially give buyers, at least those we load first, some extra free time to load resources,.... 

Buyers load in the order in which they appear in `interestGroupBuyers`?

MattMenke2 commented:
> > That does potentially give buyers, at least those we load first, some extra free time to load resources,....
> 
> Buyers load in the order in which they appear in `interestGroupBuyers`?

Mostly.  We have a global maximum of 10 buyer processes at once (On desktop).  Each buyer origin uses a single process (across all of Chrome).  We request buyer processes from Chrome in the order buyers appear, and new processes are assigned on a first-come first-served basis.  However, if we already have a process for a buyer, due to another auction (or a different component auction within the same auction), that buyer will jump the queue, since it can share the existing process.  The other auction could be on the same tab or another tab (navigated to an entirely different site).

If there are multiple component auctions in a single top-level auction, we start loading the buyers in component auctions in the order they appear as well.

On Android, where resources are more constrained, we load all worklets in the single renderer process (except for worklets on origins that are isolated for reasons I'm not familiar with), so we don't block loading bidder worklets on process creation, and the 10 process limit does not apply.  It's a bit weird that having fewer system resources make us load scripts and allocate JavaScript memory more aggressively, but that is how things currently work out.

> > I'm not sure if which auctionConfig values can be promises has been documented yet,...
> 
> Any heads up about the Promises change would be appreciated.

https://crbug.com/1394750 is the relevant Chrome bug, though it doesn't list details, unfortunately.  We've already started landing code for this.  Some of the signals fields (auctionSignals, perBuyerSignals, sellerSignals) will be able to be promises, so a page can "start" an auction before it has all the data it needs, which can then be fetched by another request.  Once an auction is started, we can start load buyer scripts, trusted buyer signals, and seller scripts while we're waiting on all 3 of those fields.  Being able to invoke generateBid() is blocked on receiving auctionSignals and perBuyerSignals, and scoreAd() calls are blocked on sellerSignals as well.

This allows the seller JS to load separately from the signals data, and it also allows, for example, a contextual auction to be run while we're loading buyer scripts, and then the contextual winning bid to be passed to sellerSignals, so the seller can reject anything that it thinks is worse than the contextual ad.

dmdabbs commented:
>We have a global maximum of 10 buyer processes at once (On desktop). 
...
If there are multiple component auctions in a single top-level auction, we start loading the buyers in component auctions in the order they appear as well.

Is there a similar process resource limit for sellers (I expect component auctions will be the norm)?

MattMenke2 commented:
> Is there a similar process resource limit for sellers (I expect component auctions will be the norm)?

Yes.  There's a limit of 3 seller processes.  Their process reuse logic is largely the same as buyers, with the exception that in component auctions, we don't assigned a process to the top-level seller until all component sellers have received a process, to avoid deadlock.  Android again behaves differently, reusing the renderer process.

The current process limits were chosen arbitrarily - we'll doubtless want to experiment and figure out what works best, both in terms of resource usage / performance, and how FLEDGE auctions are structured in practice.

MattMenke2 commented:
> It will be critical for pubs to have access to actionable stats that will help them remediate timing issues. AdOps is already complicated.

FLEDGE auctions are potentially very parallel, which can make it difficult to explain why a particular auction was slow.  I agree that we'll want to figure out how to report performance information to buyers and sellers, as well as expose more detailed information in DevTools.

dmdabbs commented:
>Yes. There's a limit of 3 seller processes. 

Thanks. Three seems low (on the assumption that SSPs on a page today will also be component sellers.) Three makes sense since that's how many are probably actively running auctions now. 

Looks like your [UMA](https://bugs.chromium.org/p/chromium/issues/detail?id=1344690) capture that. Is there any access to these sorts of numbers outside of Google? Perhaps something you could publish quarterly along with other info for CMA.

Would be great if the resource limits and allocation details get documented somethere, perhaps in Release Notes or First OT Details.

MattMenke2 commented:
> > Yes. There's a limit of 3 seller processes.
> 
> Thanks. Three seems low (on the assumption that SSPs on a page today will also be component sellers.) Three makes sense since that's how many are probably actively running auctions now.

The limit of three predates multiple-seller auctions.  I do think it's something we should revisit, sooner rather than later.  Probably the 10 as well, though that's less urgent.  We had no idea how many buyers would typically be participating in an auction when we set them, and only expected one seller.

> Looks like your [UMA](https://bugs.chromium.org/p/chromium/issues/detail?id=1344690) capture that. Is there any access to these sorts of numbers outside of Google? Perhaps something you could publish quarterly along with other info for CMA.

Decisions on what we can share, and when, are above my pay grade, so will have to get a useful response to that proposal from someone more informed on that side of things than I.

> Would be great if the resource limits and allocation details get documented somethere, perhaps in Release Notes or First OT Details.

There's always a question of how much time to spend documenting current behavior vs trying to improve it (and thus either making the documents become outdated, or investing in updating the document yet again).  None of the current process limit logic is set in stone, apart perhaps from trying to load buyers and sellers in order.  That having been said, I think you're right that it would be useful to be more transparent about how this works.

dmdabbs commented:
Appreciate your time to respond and discuss here.
